[Bug] 无入口文件的包无法解析

**复现过程**
unveilr@2.0.0-alpha.2-win-x64.exe wx 
**预期结果**

**实际结果**
[ERROR] 11:32:53 <GlobalExceptionCaught> Error: GetZArray fail: t is not iterable
    at C:\snapshot\unveilr\dist\parser-70c91931.js:7:473887
    at new Promise (<anonymous>)
    at parserZArray (C:\snapshot\unveilr\dist\parser-70c91931.js:7:473746)
    at Object.V [as parseWxml] (C:\snapshot\unveilr\dist\parser-70c91931.js:7:484828)
    at WxmlParser.parseV1 (C:\snapshot\unveilr\dist\traverse-16b693ce.js:7:24900)
    at C:\snapshot\unveilr\dist\index.js:7:15281
    at Array.forEach (<anonymous>)
    at e (C:\snapshot\unveilr\dist\index.js:7:15215)
    at Object.run (C:\snapshot\unveilr\dist\index.js:7:15842)
    at WorkerPool.<anonymous> (C:\snapshot\unveilr\node_modules\threads\dist\master\pool.js:85:48)
**相关截图**
![image](https://user-images.githubusercontent.com/4652023/227834663-29d8e806-4d11-42b8-9f5c-35b9fa73092c.png)

**运行环境、版本信息**<br/>
 - OS:Windows 10 专业版 19045.2728
 - NodeJS:18.14.1
 - Version：unveilr@2.0.0-alpha.2-win-x64

**其他描述**
实际上并没有 C:\snapshot\路径

--- a/src/core/controller/TraverseController.ts
+++ b/src/core/controller/TraverseController.ts
@@ -5,12 +5,12 @@ import { BuildParams, traverseAST } from '@utils/ast'
 
 export class TraverseController<DataType extends object = object> {
   visitors: Visitor
-  private fileBuilder: unknown
+  private fileBuilder: BabelFileResult | ProduciblePath | BuildParams
   private readonly dataSet: DataType
   constructor(path: ProduciblePath, opt?: TraverseOptions)
   constructor(builder: BuildParams, opt?: TraverseOptions)
   constructor(file: BabelFileResult, opt?: TraverseOptions)
-  constructor(fileBuilder: unknown, visitor?: Visitor) {
+  constructor(fileBuilder: BabelFileResult | ProduciblePath | BuildParams, visitor?: Visitor) {
     this.setFileBuilder(fileBuilder)
     this.addVisitors(visitor)
     this.dataSet = Object.create(null)
@@ -22,7 +22,7 @@ export class TraverseController<DataType extends object = object> {
   }
 
   async traverse(opt?: Pick<TraverseOptions, 'scope' | 'noScope'>): Promise<void> {
-    return traverseAST(this.fileBuilder, { ...this.visitors, ...opt })
+    return traverseAST(this.fileBuilder as BuildParams, { ...this.visitors, ...opt })
   }
 
   private _mergeVisitors(...source: Visitor[]): Visitor {
@@ -44,35 +44,7 @@ export class TraverseController<DataType extends object = object> {
     return dataMap
   }
 
-  setItem(k: keyof DataType, v: DataType[keyof DataType]): this {
-    this.dataSet[k] = v
-    return this
-  }
-  changeItem(
-    k: keyof DataType,
-    handler: (v: DataType[keyof DataType]) => DataType[keyof DataType],
-    _default: DataType[keyof DataType],
-  ): this {
-    this.setItem(k, handler(this.getItem(k, _default)))
-    return this
-  }
-
-  getItem(k: keyof DataType, _default: DataType[keyof DataType]): DataType[keyof DataType] {
-    return this.dataSet[k] || _default
-  }
-  popItem(k: keyof DataType, _default: DataType[keyof DataType]): DataType[keyof DataType] {
-    const data = this.dataSet[k]
-    delete this.dataSet[k]
-    return data || _default
-  }
-  get data() {
-    return this.dataSet
-  }
-
-  setFileBuilder(file: BabelFileResult): this
-  setFileBuilder(path: ProduciblePath): this
-  setFileBuilder(builder: BuildParams): this
-  setFileBuilder(v: unknown): this {
+  setFileBuilder(v: BabelFileResult | ProduciblePath | BuildParams): this {
     this.fileBuilder = v
     return this
   }

--- a/src/core/controller/WxapkgController.ts
+++ b/src/core/controller/WxapkgController.ts
@@ -7,7 +7,7 @@ import { WxmlParser } from '@core/parser/wxapkg/WxmlParser'
 import { ScriptParser } from '@core/parser/wxapkg/ScriptParser'
 import { AppConfigParser } from '@core/parser/wxapkg/AppConfigParser'
 import { matchScripts } from '@utils/matchScripts'
-import { ParserLike, TraverseVisitorKeys } from '@core/parser/wxapkg/types'
+import { ParserLike, TraverseVisitorKeys, TVSubjectType } from '@core/parser/wxapkg/types'
 import { createExposed, traverseModule } from '@core/workers/traverse'
 import { Saver } from '@utils/classes/Saver'
 import { WorkerController } from '@core/controller/WorkerController'
@@ -15,6 +15,7 @@ import { unlink } from '@utils/unlink'
 import { BaseError } from '@utils/exceptions'
 import { getConfig, getInnerConfig } from '@core/controller/ConfigController'
 import { getWccVersion } from '@utils/getWccVersion'
+import { Observable } from 'observable-fns'
 
 export class WxapkgError extends BaseError {}
 export type ParsersKey = TraverseVisitorKeys | 'WxmlParserV1'
@@ -44,7 +45,7 @@ export class WxapkgDecompiler extends BaseLogger {
   // 解包器
   private readonly extractor: WxapkgExtractor
   // 解析器
-  readonly parsers: Record<ParsersKey, ParserLike>
+  readonly parsers: Record<ParsersKey, ParserLike<Observable<TVSubjectType>, Promise<void>>>
   // 解析后使用的保存器
   private readonly saver: Saver
   // 遍历列表
@@ -67,6 +68,7 @@ export class WxapkgDecompiler extends BaseLogger {
     this.checkTypeAvailable()
     return (
       this.type === WxapkgType.APP_V3 ||
+      this.type === WxapkgType.APP_V4 ||
       this.type === WxapkgType.APP_SUBPACKAGE_V2 ||
       this.type === WxapkgType.APP_PLUGIN_V1
     )
@@ -87,12 +89,17 @@ export class WxapkgDecompiler extends BaseLogger {
       this.type === WxapkgType.APP_V1 ||
       this.type === WxapkgType.APP_V2 ||
       this.type === WxapkgType.APP_V3 ||
+      this.type === WxapkgType.APP_V4 ||
       this.type === WxapkgType.GAME
     )
   }
   // 是否是小程序插件
   get isAppPlugin() {
     return this.type === WxapkgType.APP_PLUGIN_V1
+  }
+  // 是否是游戏插件
+  get isGamePlugin() {
+    return this.type === WxapkgType.GAME_PLUGIN
   }
   // 保存的目录
   get saveDirectory(): PathController {
@@ -149,6 +156,7 @@ export class WxapkgDecompiler extends BaseLogger {
       case WxapkgType.APP_V1:
       case WxapkgType.APP_V2:
       case WxapkgType.APP_V3:
+      case WxapkgType.APP_V4:
       case WxapkgType.APP_SUBPACKAGE_V1:
       case WxapkgType.APP_SUBPACKAGE_V2:
       case WxapkgType.APP_PLUGIN_V1:
@@ -178,7 +186,7 @@ export class WxapkgDecompiler extends BaseLogger {
   async initTraverseList() {
     this.checkIsSaveExtracted()
     // 源码路径
-    const bdc = this.extractor.getSourceDir()
+    const bdc = this.extractor.getSourceDir(this.isAppPlugin || this.isGamePlugin)
     const setApp = async (options?: SetAppOptions) => {
       const {
         serviceSource = await bdc.join(WxapkgKeyFile.APP_SERVICE).read('utf8'),
@@ -233,6 +241,7 @@ export class WxapkgDecompiler extends BaseLogger {
 
     switch (this.type) {
       case WxapkgType.APP_V1:
+      case WxapkgType.APP_V4:
         await setApp({
           viewSource: matchScripts(await bdc.join(WxapkgKeyFile.PAGE_FRAME_HTML).read('utf8')),
         })
@@ -308,6 +317,8 @@ export class WxapkgDecompiler extends BaseLogger {
 export class WxapkgController extends BaseLogger {
   private readonly decompilers: WxapkgDecompiler[]
   private readonly mainSaveDir: PathController
+  // 解析结果的Promise
+  private readonly parsersPromise: Promise<void>[]
   constructor(wxapkgList: ProduciblePath[])
   constructor(...wxapkgList: ProduciblePath[])
   constructor(options: WxapkgControllerOptions)
@@ -324,6 +335,7 @@ export class WxapkgController extends BaseLogger {
       wxAppId && decompiler.setExtractorWxAppId(wxAppId)
       return decompiler
     })
+    this.parsersPromise = []
   }
 
   async calcDecompilersSaveDir() {
@@ -353,7 +365,7 @@ export class WxapkgController extends BaseLogger {
   }
 
   async invokeTraverseWorker(): Promise<void> {
-    function forEachVisitors(decompiler: WxapkgDecompiler, visitors: ParsersKey[], exposed: TraverseExposed) {
+    const forEachVisitors = (decompiler: WxapkgDecompiler, visitors: ParsersKey[], exposed: TraverseExposed) => {
       const seen = new Set()
       visitors.forEach((key) => {
         if (key === 'WxmlParserV1') {
@@ -366,12 +378,14 @@ export class WxapkgController extends BaseLogger {
         }
         const parserLike = decompiler.parsers[key]
         exposed.setVisitor(key)
-        !seen.has(parserLike) && parserLike.parse(exposed.observable())
+        !seen.has(parserLike) && this.parsersPromise.push(parserLike.parse(exposed.observable()))
         seen.add(parserLike)
       })
     }
     const results = await Promise.all(this.decompilers.map((d) => d.makeParserTraverse()))
-    const tasks = results.flat().filter(Boolean)
+    const temp: TraverseData[] = []
+    results.forEach((items) => temp.push(...items))
+    const tasks = temp.filter(Boolean)
     if (!tasks.length) {
       this.logger.warn('No task to traverse')
       return
@@ -396,6 +410,7 @@ export class WxapkgController extends BaseLogger {
   }
 
   async saveFiles() {
+    await Promise.all(this.parsersPromise)
     await Promise.all(this.decompilers.map((d) => d.save()))
   }
 

--- a/src/core/extractor/WxapkgExtractor.ts
+++ b/src/core/extractor/WxapkgExtractor.ts
@@ -114,7 +114,7 @@ export class WxapkgExtractor extends BaseExtractor {
   get saveDirectory() {
     return this.saver.saveDirectory
   }
-  getSourceDir(): PathController {
+  getSourceDir(isPlugin: boolean): PathController {
     const result = this.saver.saveDirectory
       .join(this.sourcePath || '.')
       .reload()
@@ -125,7 +125,7 @@ export class WxapkgExtractor extends BaseExtractor {
           baseName === WxapkgKeyFile.APP_SERVICE ||
           baseName === WxapkgKeyFile.APPSERVICE ||
           baseName === WxapkgKeyFile.GAME ||
-          baseName === WxapkgKeyFile.PLUGIN_JSON
+          (isPlugin && baseName === WxapkgKeyFile.PLUGIN_JSON)
         )
       })
     if (!result) ExtractorError.throw(`Source code path not found, may not be a supported package`)

--- a/src/core/parser/wxapkg/AppConfigParser.ts
+++ b/src/core/parser/wxapkg/AppConfigParser.ts
@@ -7,10 +7,17 @@ import { Saver } from '@utils/classes/Saver'
 import { filter } from 'observable-fns'
 import { md5 } from '@utils/crypto'
 
-interface pageInfo {
+interface PageInfo {
   [key: string]: {
     window: { usingComponents: { [key: string]: unknown }; [key: string]: unknown }
   }
+}
+interface TabBarItem {
+  iconPath?: string
+  iconData?: string
+  selectedIconPath?: string
+  selectedIconData?: string
+  pagePath?: string
 }
 export class AppConfigParser extends BaseParser {
   private serviceSource: string
@@ -72,15 +79,18 @@ export class AppConfigParser extends BaseParser {
       }
       // tabBar
       const tabBar = config.pop('tabBar')
-      const ignoreSuffixes = 'html,wxss,json'
+      const ignoreSuffixes = 'wxml,wxs,wxss,html,json'
       if (tabBar && Array.isArray(tabBar.list)) {
-        const hashMap: { [key: string]: string }[] = Object.create(null)
-        dirCtrl.deepListDir().forEach((p) => {
-          const pCtrl = PathController.unix(p)
-          if (ignoreSuffixes.includes(pCtrl.suffixWithout)) return
-          hashMap[md5(pCtrl.readSync())] = pCtrl.crop(dirCtrl).unixpath
-        })
-        tabBar.list.forEach((item) => {
+        const hashMap: Record<string, string> = Object.create(null)
+        dirCtrl
+          .reload()
+          .deepListDir()
+          .forEach((p) => {
+            const pCtrl = PathController.unix(p)
+            if (ignoreSuffixes.includes(pCtrl.suffixWithout)) return
+            hashMap[md5(pCtrl.readSync())] = pCtrl.crop(dirCtrl).unixpath
+          })
+        tabBar.list.forEach((item: TabBarItem) => {
           item.pagePath = PathController.make(item.pagePath).whitout().unixpath
           if (item.iconData) {
             const path = hashMap[md5(item.iconData, true)]
@@ -99,7 +109,7 @@ export class AppConfigParser extends BaseParser {
         })
       }
       // usingComponents
-      const page: pageInfo = config.pop('page')
+      const page: PageInfo = config.pop('page')
       config.pop('renderer')
       Object.keys(page).forEach((key) => {
         const usingComponents = page[key].window.usingComponents
@@ -121,14 +131,22 @@ export class AppConfigParser extends BaseParser {
       save(WxapkgKeyFile.APP_JSON, result)
       // usingComponents -> json
       if (!this.serviceSource) ParserError.throw(`Service source not found!`)
-      observable.pipe<S2Observable<AppConfigServiceSubject>>(filter((v) => v.AppConfigService)).subscribe((value) => {
-        Object.entries(value.AppConfigService).forEach((args) => save(...args))
+      let resolve
+      const promise = new Promise<void>((_resolve) => (resolve = _resolve))
+      observable.pipe<S2Observable<AppConfigServiceSubject>>(filter((v) => v.AppConfigService)).subscribe({
+        next: (value) => {
+          Object.entries(value.AppConfigService).forEach((args) => save(...args))
+        },
+        complete() {
+          Object.keys(page).forEach((key) => {
+            let pCtrl = PathController.make(key)
+            if (pCtrl.suffix !== '.json') pCtrl = pCtrl.whitout('.json')
+            save(pCtrl, page[key]['window'])
+          })
+          resolve && resolve()
+        },
       })
-      Object.keys(page).forEach((key) => {
-        let pCtrl = PathController.make(key)
-        if (pCtrl.suffix !== '.json') pCtrl = pCtrl.whitout('.json')
-        save(pCtrl, page[key]['window'])
-      })
+      return promise
     } catch (e) {
       ParserError.throw('Parse failed! ' + e.message)
     }

--- a/src/core/parser/wxapkg/ScriptParser.ts
+++ b/src/core/parser/wxapkg/ScriptParser.ts
@@ -4,15 +4,31 @@ import { reformat } from '@utils/reformat'
 import { S2Observable, ScriptParserSubject, TVSubject } from '@core/parser/wxapkg/types'
 import { Saver } from '@utils/classes/Saver'
 import { filter } from 'observable-fns'
+import { PathController } from '@core/controller/PathController'
 
 export class ScriptParser extends BaseParser {
   constructor(saver: Saver) {
     super(saver)
   }
   async parse(observable: S2Observable<TVSubject>): Promise<void> {
-    observable.pipe<S2Observable<ScriptParserSubject>>(filter((v) => v.ScriptParser)).subscribe((value) => {
-      Object.entries(value.ScriptParser).forEach(([path, buffer]) => this.saver.add({ path, buffer }))
+    let resolve
+    const promise = new Promise<void>((_resolve) => (resolve = _resolve))
+    observable.pipe<S2Observable<ScriptParserSubject>>(filter((v) => v.ScriptParser)).subscribe({
+      next: (value) => {
+        Object.entries(value.ScriptParser).forEach(([path, buffer]) => {
+          if (path.startsWith('__plugin__')) {
+            if (!PathController.make(path).suffixWithout) {
+              path = '__plugin__/index.js'
+            }
+          }
+          this.saver.add({ path, buffer })
+        })
+      },
+      complete() {
+        resolve && resolve()
+      },
     })
+    return promise
   }
 
   static visitor(subject: ScriptParserSubject): Visitor {
@@ -22,7 +38,7 @@ export class ScriptParser extends BaseParser {
         if (callee.type === 'Identifier' && callee.name === 'define') {
           const args = path.get('arguments')
           const [filenamePathNode, sourcePathNode] = args
-          if (filenamePathNode.node.type !== 'StringLiteral') return
+          if (!filenamePathNode.isStringLiteral()) return
           if (sourcePathNode.node.type !== 'FunctionExpression') return
           const filename = filenamePathNode.node.value
           const body = sourcePathNode.get('body.body')

--- a/src/core/parser/wxapkg/WxmlParser.ts
+++ b/src/core/parser/wxapkg/WxmlParser.ts
@@ -70,18 +70,26 @@ export class WxmlParser extends BaseParser {
     super(saver)
   }
   async parse(observable: S2Observable<TVSubject>): Promise<void> {
+    let resolve
+    const promise = new Promise<void>((_resolve) => (resolve = _resolve))
     // 订阅wxml解析器
-    observable.pipe<S2Observable<WxmlParserV3Subject>>(filter((v) => v.WxmlParserV3)).subscribe((value) => {
-      const data = value.WxmlParserV3
-      if (!data) return
-      const { code, json, z } = data
-      const dir = this.dir
-      parseWxml(code, dir, json, z).then((result) => {
-        Object.entries(result).forEach(([path, buffer]) => {
-          this.saver.add({ path, buffer })
+    observable.pipe<S2Observable<WxmlParserV3Subject>>(filter((v) => v.WxmlParserV3)).subscribe({
+      next: (value) => {
+        const data = value.WxmlParserV3
+        if (!data) return
+        const { code, json, z } = data
+        const dir = this.dir
+        parseWxml(code, dir, json, z).then((result) => {
+          Object.entries(result).forEach(([path, buffer]) => {
+            this.saver.add({ path, buffer })
+          })
         })
-      })
+      },
+      complete() {
+        resolve && resolve()
+      },
     })
+    return promise
   }
 
   async parseV1() {

--- a/src/core/parser/wxapkg/WxssParser.ts
+++ b/src/core/parser/wxapkg/WxssParser.ts
@@ -87,18 +87,30 @@ export class WxssParser extends BaseParser {
     const addSaver = (data: Dict) => {
       Object.entries(data).forEach(([path, buffer]) => this.saver.add({ path, buffer }))
     }
+    let resolve
+    const promise = new Promise<void>((_resolve) => (resolve = _resolve))
     // 订阅公共样式
-    observable
-      .pipe<S2Observable<WxssParserCommonSubject>>(filter((v) => v.WxssParserCommon))
-      .subscribe((data) => addSaver(data.WxssParserCommon))
+    observable.pipe<S2Observable<WxssParserCommonSubject>>(filter((v) => v.WxssParserCommon)).subscribe({
+      next: (data) => addSaver(data.WxssParserCommon),
+      complete() {
+        resolve && resolve()
+      },
+    })
     // 订阅公共样式2 _C
-    observable
-      .pipe<S2Observable<WxssParserCommon2Subject>>(filter((v) => v.WxssParserCommon2))
-      .subscribe((data) => addSaver(data.WxssParserCommon2))
+    observable.pipe<S2Observable<WxssParserCommon2Subject>>(filter((v) => v.WxssParserCommon2)).subscribe({
+      next: (data) => addSaver(data.WxssParserCommon2),
+      complete() {
+        resolve && resolve()
+      },
+    })
     // 订阅非公共样式
-    observable
-      .pipe<S2Observable<WxssParserSubject>>(filter((v) => v.WxssParser))
-      .subscribe((data) => addSaver(data.WxssParser))
+    observable.pipe<S2Observable<WxssParserSubject>>(filter((v) => v.WxssParser)).subscribe({
+      next: (data) => addSaver(data.WxssParser),
+      complete() {
+        resolve && resolve()
+      },
+    })
+    return promise
   }
   //  读取 `setCssToHead` 函数
   static visitorSetCssToHead(subject: WxssParserSubject): Visitor {

--- a/src/core/workers/traverse.ts
+++ b/src/core/workers/traverse.ts
@@ -29,7 +29,8 @@ export function createExposed() {
     async startTraverse(code: string) {
       const tCtrl = new TraverseController({ code })
       tCtrl.addVisitors(...Object.values(visitors))
-      return tCtrl.traverse()
+      await tCtrl.traverse()
+      subject.complete()
     },
     setVisitor(...keys: Array<keyof TraverseVisitorMap>) {
       keys.forEach((key) => (visitors[key] = visitorsFn[key](subject)))

--- a/src/enum/Wxapkg.ts
+++ b/src/enum/Wxapkg.ts
@@ -27,6 +27,7 @@ export enum WxapkgType {
   APP_V1 = 'APP_V1',
   APP_V2 = 'APP_V2',
   APP_V3 = 'APP_V3',
+  APP_V4 = 'APP_V4',
   /** 对标 {@link APP_V1},{@link APP_V2}*/
   APP_SUBPACKAGE_V1 = 'APP_SUBPACKAGE_V1',
   /** 对标 {@link APP_V3}*/

--- a/src/utils/checkWxapkg.ts
+++ b/src/utils/checkWxapkg.ts
@@ -27,8 +27,11 @@ export async function checkWxapkgType(v: ProduciblePath | string[]): Promise<Wxa
     fileList = v as string[]
   }
   if (fileList.every((filename) => filename.startsWith('WA'))) return WxapkgType.FRAMEWORK
-  // APP_V1
-  if (fileList.includes(WxapkgKeyFile.PAGE_FRAME_HTML)) return WxapkgType.APP_V1
+  // APP_V1/APP_V4
+  if (fileList.includes(WxapkgKeyFile.PAGE_FRAME_HTML)) {
+    if (fileList.includes(WxapkgKeyFile.COMMON_APP)) return WxapkgType.APP_V4
+    return WxapkgType.APP_V1
+  }
   // APP_V3/APP_SUBPACKAGE_V2
   if (fileList.includes(WxapkgKeyFile.COMMON_APP))
     return fileList.includes(WxapkgKeyFile.APP_WXSS) ? WxapkgType.APP_V3 : WxapkgType.APP_SUBPACKAGE_V2

--- a/src/utils/exceptions.ts
+++ b/src/utils/exceptions.ts
@@ -2,7 +2,7 @@ import { BaseLogger } from '@utils/logger'
 
 class GlobalExceptionCaught extends BaseLogger {
   constructor() {
-    super()
+    super('Exception')
     const ignores = 'exit,cancel,pass'.split(',')
     const handler = (e: Error | string) => (ignores.includes(String(e)) ? '' : this.logger.error(e))
     process.on('uncaughtException', handler)

